cmd/tealdbg/debugger.go

```diff
@@ -181,7 +184,54 @@ func (s *session) Step() {
 	func() {
 		s.mu.Lock()
 		defer s.mu.Unlock()
-		s.debugConfig = debugConfig{BreakAtLine: stepBreak}
+		s.debugConfig = debugConfig{StepBreak: true}
+	}()
+
+	s.resume()
+}
+
+func (s *session) StepOver() {
+	// Get the first TEAL opcode in the line
+	currentOp := strings.Fields(s.lines[s.line.Load()])[0]
```

Comment:
@algochoi Checking to see if you're aware of the rationale:  How is it safe to access `s.line` outside of `s.mu`?

* I understand the value uses atomics to provide _safe_ reads/writes.
* The concern is:  state changes for `line` are liable to happen in a way that's out-of-sync with the rest of `session` state.
  * It seems to me `session` state is intended to be acted upon transactionally.  For example, I presume that's why `Control` implementations here wrap invocations in an anonymous function.
* Provided the understanding is accurate _and_ if no other explanation exists, I suggest moving all `s.line` reads/writes within `s.mu` control.
  * I think additional attention is needed in `Debugger.Update` because `d.getSession` uses the mutex.  And from a quick read, it seems `s.mu` is _not_ re-entrant.

Reply:
I agree moving `s.line` into the lock is safer - pushed the changes.

Might be a pedantic comment - I don't *think* `s.line` would change unless `Update` is unblocked, and under current circumstances, I can't think of how `Update` would be unblocked in the middle of `StepOver` - I think `Update` would only resume  some time after `s.resume()` sends the `ack` to the channel. Of course, this doesn't change the fact that it's still safer to access that state inside a lock, so I agree with you.

Reply:
@algochoi Thanks for the change.  However, I am thinking more broadly.  What are your thoughts on the note below?

I can also discuss live if the note is too involved.

As a motivating example:  Consider a scenario where the user issues a command (e.g. step) after `s.line.Store(state.Line)` in `Debugger.Update`, but _before_ updating the remaining state in `go func(localState logic.DebugState)`.

* Such an example breaks the implied transactional update model.  In the example, `s.line` points to an updated line, but other state hasn't been updated yet.
* And even if we can convince ourselves that the scenario works today, I don't understand why we'd take the chance.  It feels to me like it'd be more robust to ensure mutual exclusion for _all_ state changes (or otherwise change the model).  
* Extending the thinking - I don't understand why `s.line` receives special treatment via atomics.  If mutual exclusion is enforced for _all_ state changes, then there ought to be _no_ visibility concerns.

I realize some (or possibly all) of these concerns predate the PR.  I'm at least trying to gauge if we agree on the problem scope.  Because _if_ we agree, it feels like there's a fundamental issue here. 
 And we can then assess when/where to take action.

Reply:
First reason is `line` is really get used by debugger and part of its state. The `state` is not really needed and the debugger here is a "persistence" place to have a an initial state and apply changes as the evaluation goes.
But I agree, synchronous update would not hurt.

Reply:
@algochoi @algorandskiy Thanks for the discussion here to help me understand the space.  I think we're good to resolve.  

I opened https://github.com/algorand/go-algorand/issues/3850 to track the concern.  We can follow up out-of-band to the PR.